src/iterator/backend.rs

```diff
@@ -414,19 +414,20 @@ pub enum PollResult<O> {
 }
 
 /// An infinite iterator of received signals.
-pub struct SignalIterator<R, E: Exfiltrator> {
-    signals: SignalDelivery<R, E>,
+pub struct SignalIterator<SD, E: Exfiltrator> {
+    signals: SD,
     iter: Pending<E>,
 }
 
-impl<R, E: Exfiltrator> SignalIterator<R, E>
-where
-    R: 'static + AsRawFd + Send + Sync,
-{
+impl<SD, E: Exfiltrator> SignalIterator<SD, E> {
     /// Create a new infinite iterator for signals registered with the passed
     /// in [`SignalDelivery`] instance.
-    pub fn new(mut signals: SignalDelivery<R, E>) -> Self {
-        let iter = signals.pending();
+    pub fn new<R>(mut signals: SD) -> Self
+    where
+        SD: BorrowMut<SignalDelivery<R, E>>,
+        R: 'static + AsRawFd + Send + Sync,
```

Comment:
I feel a bit uneasy about these trait bounds, for two reasons:

* I think `R` and the bounds should be on the impl block (if there are some methods that aren't influenced by it, then maybe having two impl blocks). Because `new` or `handle` or anything else is not influenced by `R`, really, so they should not have it in their signatures.
* Are we sure there's no situation at which the user might need to explicitly specify `R`?

Reply:
Moving the trait bounds to the impl isn't possible in the current state because this woudln't constrain the R parameter. I tried this first but found out it's not working. This issue is desribed [here](https://github.com/rust-lang/rust/issues/56209).

It would be possible to introducing a R parameter directly in the struct:

```
struct SignalIterator<R, SD, E>
where
    R: 'static + AsRawFd + Send + Sync,
    E: Exfiltrator,
    SD: BorrowMut<SignalDelivery<R, E>,
{
    std::marker::PhantomData<R>,
    ...
}

impl<R, SD, E> SignalIterator<R, SD, E>
where
    R: 'static + AsRawFd + Send + Sync,
    E: Exfiltrator,
    SD: BorrowMut<SignalDelivery<R, E>,
{
    ...
}
```

Then the R parameter is constraint in the impl and it is possible to add the other bounds there, too.

Can you elaborate your second point? I'm not sure if you are concerned about the missing possibility to explicitly specify it by the user or the user has to specify it in some rare cases.

Reply:
I was worried that the user would have to specify it in some rare cases (or less rare ones).

But on the second look, this is the backend of the iterator and the only ones interacting with it are the authors of the async crates, so even if it was true, it would be fine. So let's leave it as it is.